## SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

### Выполним подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Исследуем таблицы - выведем первые строки

In [2]:
books = "SELECT * FROM books"
books = pd.io.sql.read_sql(books, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
authors = "SELECT * FROM authors"
authors = pd.io.sql.read_sql(authors, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
publishers = "SELECT * FROM publishers"
publishers = pd.io.sql.read_sql(publishers, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
ratings = "SELECT * FROM ratings"
ratings = pd.io.sql.read_sql(ratings, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
reviews = "SELECT * FROM reviews"
reviews = pd.io.sql.read_sql(reviews, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:
tables = [books, authors, publishers, ratings, reviews]

In [8]:
for i in tables:
    i.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        

### Задания

#### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [9]:
query = "SELECT DISTINCT(COUNT(book_id)) FROM books WHERE publication_date > '2000-01-01'"

In [10]:
pd.io.sql.read_sql(query, con = engine)

count
0    819

#### Вывод:
После 1 января 2000 года выпущено 819 книг.

#### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [21]:
query = '''SELECT b.title, COUNT(DISTINCT(rw.text)) AS "Кол-во обзоров", AVG(r.rating) AS "Средняя оценка"
           FROM ratings r
           RIGHT JOIN books b ON r.book_id = b.book_id
           LEFT JOIN reviews rw ON b.book_id = rw.book_id
           GROUP BY b.book_id
           ORDER BY "Кол-во обзоров" DESC, "Средняя оценка" DESC'''




In [12]:
pd.io.sql.read_sql(query, con = engine)

title  Кол-во обзоров  \
0                              Twilight (Twilight  #1)               7   
1    Harry Potter and the Prisoner of Azkaban (Harr...               6   
2    Harry Potter and the Chamber of Secrets (Harry...               6   
3                                       The Book Thief               6   
4                                     The Glass Castle               6   
..                                                 ...             ...   
995  Disney's Beauty and the Beast (A Little Golden...               0   
996                               Leonardo's Notebooks               0   
997                          Essential Tales and Poems               0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel               0   
999                            The Natural Way to Draw               0   

     Средняя оценка  
0          3.662500  
1          4.414634  
2          4.287500  
3          4.264151  
4          4.206897  
..              ...  
995        4.000000  
996        4.000000  
997        4.000000  
998        3.666667  
999        3.000000  

[1000 rows x 3 columns]

#### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [13]:
query = '''with a as(SELECT b.title, b.num_pages, p.publisher
           FROM books b
           JOIN publishers p ON b.publisher_id = p.publisher_id
           WHERE b.num_pages > 50)
           SELECT a.publisher, COUNT(title)
           FROM a
           GROUP BY publisher
           ORDER BY COUNT(title) DESC
           LIMIT 1'''

In [14]:
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

#### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [15]:
query = '''with q as(SELECT book_id, COUNT(rating) as q_count
           FROM ratings
           GROUP BY book_id
           HAVING COUNT(rating) >= 50)
           SELECT a.author, AVG(r.rating)
           FROM q
           JOIN books b ON q.book_id = b.book_id
           JOIN authors a ON b.author_id = a.author_id
           JOIN ratings r ON b.book_id = r.book_id
           GROUP BY a.author
           ORDER BY AVG(r.rating) DESC
           LIMIT 1'''

In [16]:
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

#### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
query = '''with w as (with q as (SELECT username, COUNT(rating_id)
           FROM ratings
           GROUP BY username
           HAVING COUNT(rating_id) > 50)
           SELECT q.username, COUNT(review_id) as count_rw
           FROM q
           JOIN reviews r ON q.username = r.username
           GROUP BY q.username)
           SELECT AVG(w.count_rw)
           FROM w'''

In [18]:
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

### Общие выводы:
- После 1 января 2000 года было выпущено 819 книг из 1000 книг, собранных в базе.
- Наибольшее количество обзоров у книги Twilight (Twilight #1), при этом оценка 3.66. Затем по количеству обзоров идет книга Harry Potter and the Prisoner of Azkaban, и рейтинг у нее выше - 4.41.
- Наибольшее количество книг (42), толщиной более 50 страниц, выпускает издательство Penguin Books.
- Автор с самой высокой средней оценкой (4.29) книг, которых оценили более 50 раз - J.K. Rowling/Mary GrandPré.
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33.